### DISCUSSION TITLE
Why do the parameters of a network is changing when i make gradient =0?

### DISCUSSION THREAD

**Discussion by maobenz** [jax-ml/jax#10224] [Q&A]
I use the stop_gradient() to make some paramters of a network be 0, and i print it out and it proves they are all zero.

However, after some iterations, the value of parameters is changing.

Code:

 ```
Class A:
     Class B:
     Class C:


    def _call_:
         b = B(input)
         c = C(b)
         c = stop_gradient(c)
```

```
print(grad)
## grad of model c is all zero
print(state.optimizer.target['model']['c'])
## after one iterations
the value of the paramter is changing

```
Why?? is the reason that the value of parameters of model b is changing??
How do i really stop model c learn??
Thanks

**Comment by YouJiacheng** [comment]
What optimizer did you use?

**Reply by maobenz** [reply]
Thanks! flax.optim.Optimizer
The problem is that when i stop the C model, and it seems like B model doesn't work as well but it should be. And the C model's parameters are not really stopped.
However, when i stop B model, B model's parameters stop to change.

**Reply by YouJiacheng** [reply]
I mean the type of optimizer, SGD with momentum, Adam, etc.

**Reply by YouJiacheng** [reply]
If you stop grad for C model output, B model won't receive gradient as well.

**Reply by maobenz** [reply]
Thanks.
Actually i find if i stop C model's output's gradient, B model's parameters have also grad, which is not zero and the parameters are changing.
the relationship of B and C is:
b = B(input)
c = C(b,d,e) # d and e are other varibles, only b is the output of B model output.
And B and C are created in model A.

And optimizer is adam optimizer.

**Reply by maobenz** [reply]
Sorry, maybe i should say it more clearly.
i made some experiments
My archetecture is like this
```
Class A:

    def setup():
        self.B = B()
        self.C = C()
   
    def __call__(x1,x2,x3):  # a b c are input
        y= self.B(x1)
        z = self.C(y,x2,x3)

# below is the defination of model B and C
Class B:

Class C:


def main():
     model = A()
     result = model.apply(params)

```
Firstly,
I use the params = jax.lax.stop_gradient(params)
```
def main():
     model = A()
     params = jax.lax.stop_gradient(params)
     result = model.apply(params)
```
And all gradients of the A model are zero and the whole model stops to learn.

Secondly,  
```
def __call__(x1,x2,x3):  # a b c are input
    y= self.B(x1)
    y = jax.stop_gradient(y)
    z = self.C(y,x2,x3)
```
and B model's gradient is all zero and stops to learn 
C model's parameters are changing and learning and gradients are not zero.

Thirdly
```
def __call__(x1,x2,x3):  # a b c are input
    y= self.B(x1)
    z = self.C(y,x2,x3)
    z = jax.stop_gradient(z)
```
However, C model's gradients are zero, but the parameters are changing.
B model's gradients are not zero and parameters are changing.

So it is my question.
It seems like when i use the stop_gradient ot model c and gradients are zero but it is learning.

**Reply by maobenz** [reply]
and i think optimizer.target['model'] is  all paramters of the model A . I print it up and i also print the jax.value_and_grad to check the model's gradient.

**Reply by YouJiacheng** [reply]
I can't clearly figure out the problem, but at least you can manually copy the parameters of model C from the old state.
`flax.optim` is deprecated and I don't know how to work with it. If you use `optax` I can offer some help.
see https://github.com/google/flax/blob/main/docs/flip/1009-optimizer-api.md

**Reply by maobenz** [reply]
Thanks. i will try the optax. Actually maybe i can't  manually copy the paramters because when i get the parameters from the optimizer, it is the frozen dict type and it can't be changed.

**Reply by YouJiacheng** [reply]
There are `unfrozen` and `copy`(with replace or add) APIs for frozen dict.

**Reply by maobenz** [reply]
Thanks!
I think unfrozen and copy api just create new dictionary but not change the orignal dict. And the orginal dict is also frozen.
just like 
dic = dic.copy(params)
it doesn't work

**Reply by YouJiacheng** [reply]
Yes it creates a new dictionary, but you don't need to modify the original dict! You only need to modify the optimizer state!
